bigbluebutton/greenlight
# Add Sign In

## Description

This PR implements 3 features/mechanics that arose from the Sign In process: 

- the sign in form (SignIn component)
- the server side api hooks for signing in (Sessions controller)
- the current user api hooks and display (CurrentUser, AuthContext component and Sessions helper)


**Sign In Form**
User enter email and password in Sign In form.
If email and password correspond to a user in the server database, user is redirected to Home page on Form submit.

**Server Side API Hooks**
Server checks if User exists.
Server checks if User and Password match via authenticate method.
If success, return success status, signed_in as true and store User id in session.
If error, return error status and signed_in as false.

**Current User**
Frontend:
If signed in, the Current User information is displayed (name, email, provider, signed_in?).
The currentUser object is accessible to any component via the useAuth() method import.
Backend:
The current_user helper method can now be accessed anywhere in server side to display current user.


## How To Access Current User in a Component

The whole app tree has been wrapped inside the AuthProvider component.

As per AuthContext.tsx (in components/sessions), every AuthProvider children will inherit the currentUser value/props. 

![image](https://user-images.githubusercontent.com/43917914/160632805-f5e3372d-d437-4130-b8bd-e29a41538bc3.png)

To use the currentUser value/props inside a component, import the context ` import { useAuth } from '../sessions/AuthContext';` and use the custom hook useAuth() `const currentUser = useAuth();` 

The object returned by useAuth() now contains the current_user data, as per the API hook set at /api/v1/sessions/signed_in (SessionsController).


## Testing Steps

1. Navigate to /sign_in

![image](https://user-images.githubusercontent.com/43917914/160180121-149d4f11-c44b-485c-b576-7f4be134e266.png)

2. Fill in the email and password input as an existing User in your local database
If the authentication is successful, you will be redirected to the home component.
Also, the user information is displayed to confirm the persistence of the session.

![image](https://user-images.githubusercontent.com/43917914/160180390-00b034d0-2651-487f-8ac2-916ed6ae0607.png)



## TODO:
- Clean up code

  1. createSession() api call does not belong in Form/SignIn component.
  2. fetchCurrentUser() api call does not belong in AuthContext component.
  3. review variables names with team.
  4. use Amir template for Sign In form.

- Use axios instead of fetch for api calls
- Add Alerts if Sign In is success or error
- Use Typescript
